packages/expo-router/src/fork/getStateFromPath.ts

```diff
@@ -195,65 +195,50 @@ function assertConfigDuplicates(configs: RouteConfig[]) {
+function sortConfigs(
+  aConfig: RouteConfig,
+  bConfig: RouteConfig,
+  previousSegments: string[] = []
+): number {
   // Sort config so that:
   // - the most exhaustive ones are always at the beginning
   // - patterns with wildcard are always at the end
+  const a = getConfigMetadata(aConfig);
+  const b = getConfigMetadata(bConfig);
 
   // If 2 patterns are same, move the one with less route names up
   // This is an error state, so it's only useful for consistent error messages
   if (a.pattern === b.pattern) {
-    return b.routeNames.join('>').localeCompare(a.routeNames.join('>'));
+    return a.routeNames.join('>').localeCompare(b.routeNames.join('>'));
   }
 
-  // If one of the patterns starts with the other, it's more exhaustive
-  // So move it up
-  if (
-    a.pattern.startsWith(b.pattern) &&
-    // NOTE(EvanBacon): This is a hack to make sure that `*` is always at the end
-    b.screen !== 'index'
-  ) {
+  /*
+   * If one of the patterns starts with the other, it is earlier in the config sorting.
+   * However, configs are a mix of route configs and layout configs
+   * e.g There will be a config for `/(group)`, but maybe there isn't a `/(group)/index.tsx`
+   *
+   * This is because you can navigate to a directory and its navigator will determine the route
+   * These routes should be later in the config sorting, as their patterns are very open
+   * and will prevent routes from being matched
+   *
+   * Therefore before we compare segment parts, we force these layout configs later in the sorting
+   *
+   * NOTE(marklawlor): Is this a feature we want? I'm unsure if this is a gimmick or a feature.
+   */
+  if (a.pattern.startsWith(b.pattern) && !b.isIndex) {
     return -1;
   }
 
-  if (b.pattern.startsWith(a.pattern) && a.screen !== 'index') {
+  if (b.pattern.startsWith(a.pattern) && !a.isIndex) {
     return 1;
   }
 
-  // NOTE(EvanBacon): Here we append `index` if the screen was `index` so the length is the same
-  // as a slug or wildcard when nested more than one level deep.
-  // This is so we can compare the length of the pattern, e.g. `foo/*` > `foo` vs `*` < ``.
-  const aParts = a.pattern
-    .split('/')
-    // Strip out group names to ensure they don't affect the priority.
-    .filter((part) => matchGroupName(part) == null);
-  if (a.screen === 'index' || a.screen.match(/\/index$/)) {
-    aParts.push('index');
-  }
-
-  const bParts = b.pattern.split('/').filter((part) => matchGroupName(part) == null);
-  if (b.screen === 'index' || b.screen.match(/\/index$/)) {
-    bParts.push('index');
-  }
-
-  const isAStaticRoute =
-    !a.hasChildren && // Layout configs will have children
-    !aParts.some(
-      (part) => part.startsWith(':') || part.startsWith('*') || part.includes('*not-found')
-    );
-  const isBStaticRoute =
-    !b.hasChildren &&
-    !bParts.some(
-      (part) => part.startsWith(':') || part.startsWith('*') || part.includes('*not-found')
-    );
-
```

Comment:
This code block was moved into the `getSortingMeta`. This function caches these values so they do not need to be calculated every navigation